theano/compile/nanguardmode.py

```diff
@@ -187,6 +200,47 @@ def compile_gpu_func(nan_is_error, inf_is_error, big_is_error):
             cuda_compile_failed = True
 
 
+def f_gpua_min(inp):
+    dt = inp.dtype
+    ctx_name = _name_for_ctx(inp.context)
+    key = (dt, ctx_name)
+    f = f_gpua_min.cache.get(key, None)
+    if f is None:
+        guard_in = GpuArrayType(str(dt), (False,), context_name=ctx_name)()
+        f = theano.function([guard_in], T.min(guard_in),
+                            mode='FAST_RUN', profile=False)
+        f_gpua_min.cache[key] = f
+    return f(inp)
+f_gpua_min.cache = dict()
+
+def f_gpua_max(inp):
+    dt = inp.dtype
+    ctx_name = _name_for_ctx(inp.context)
+    key = (dt, ctx_name)
+    f = f_gpua_min.cache.get(key, None)
+    if f is None:
+        guard_in = GpuArrayType(str(dt), (False,), context_name=ctx_name)()
+        f = theano.function([guard_in], T.max(guard_in),
+                            mode='FAST_RUN', profile=False)
+        f_gpua_min.cache[key] = f
+    return f(inp)
+f_gpua_max.cache = dict()
+
+
+def f_gpua_absmax(inp):
+    dt = inp.dtype
+    ctx_name = _name_for_ctx(inp.context)
+    key = (dt, ctx_name)
+    f = f_gpua_min.cache.get(key, None)
+    if f is None:
+        guard_in = GpuArrayType(str(dt), (False,), context_name=ctx_name)()
+        f = theano.function([guard_in], T.max(T.abs_(guard_in)),
+                            mode='FAST_RUN', profile=False)
+        f_gpua_min.cache[key] = f
+    return f(inp)
+f_gpua_absmax.cache = dict()
```

Comment:
You duplicate the same code 3 times... If you could refactor it to have a builder function that you just pass a lambda that does the graph to execute and include caching, it would be great, but it is optional.